File: data-structures-sorting-algorithms-advanced/src/algorithms/selectionSort.js
```javascript
/**
 * @fileoverview Implementation of the Selection Sort algorithm.
 */

const { swap } = require('../utils/arrayUtils');

/**
 * Sorts an array of numbers using the Selection Sort algorithm.
 *
 * Selection Sort works by repeatedly finding the minimum (or maximum) element
 * from the unsorted part of the array and putting it at the beginning (or end)
 * of the sorted part.
 *
 * Characteristics:
 * - Simple to understand and implement.
 * - Unstable: Yes (does not preserve the relative order of equal elements).
 *   Example: [5a, 5b, 1] -> [1, 5b, 5a]. 5b moves before 5a.
 * - In-place: Yes.
 * - Not adaptive: Performance is the same regardless of initial order.
 *
 * @param {Array<number>} arr The array of numbers to be sorted.
 * @returns {Array<number>} The sorted array. The original array is modified in-place.
 *
 * Time Complexity:
 *   - Worst-case: O(N^2)
 *   - Average-case: O(N^2)
 *   - Best-case: O(N^2) (still performs N comparisons in inner loop)
 *
 * Space Complexity:
 *   - O(1) (in-place)
 *
 * Example:
 *   selectionSort([5, 1, 4, 2, 8]) // returns [1, 2, 4, 5, 8]
 *   selectionSort([])             // returns []
 *   selectionSort([1])            // returns [1]
 */
function selectionSort(arr) {
  if (!Array.isArray(arr)) {
    throw new Error('Input must be an array.');
  }
  const n = arr.length;

  // An array with 0 or 1 element is already sorted.
  if (n <= 1) {
    return arr;
  }

  // One by one, move the boundary of the unsorted subarray.
  // The outer loop iterates from the first element up to the second to last.
  // The last element will automatically be in its correct place after n-1 passes.
  for (let i = 0; i < n - 1; i++) {
    // Find the minimum element in the unsorted portion of the array (from index i to n-1).
    let minIdx = i;
    for (let j = i + 1; j < n; j++) {
      if (arr[j] < arr[minIdx]) {
        minIdx = j;
      }
    }

    // If the minimum element is not already at the current position 'i', swap them.
    // This places the found minimum element at the correct sorted position.
    if (minIdx !== i) {
      swap(arr, i, minIdx);
    }
  }
  return arr;
}

module.exports = selectionSort;
```

File: data-structures-sorting-algorithms-enhanced/sorting_algorithms_project/algorithms/selection_sort.py
def selection_sort(arr):
    """
    Sorts an array using the Selection Sort algorithm.

    Selection Sort algorithm sorts an array by repeatedly finding the minimum element
    (considering ascending order) from the unsorted part and putting it at the beginning.
    The algorithm maintains two subarrays in a given array:
    1. The subarray which is already sorted.
    2. The remaining subarray which is unsorted.

    In every iteration of selection sort, the minimum element from the unsorted subarray
    is picked and moved to the sorted subarray.

    Args:
        arr (list): The list of comparable elements to be sorted.

    Time Complexity:
        - Worst Case: O(n^2)
        - Average Case: O(n^2)
        - Best Case: O(n^2)
        Selection sort performs (n-1) swaps and n(n-1)/2 comparisons in all cases
        (worst, average, and best). The number of swaps is minimal, which can be an
        advantage if write operations are expensive.

    Space Complexity:
        - O(1) - It sorts the array in-place, requiring no additional storage
                 proportional to the input size.

    Stability:
        - Unstable: It does not preserve the relative order of equal elements.
                    Consider `[5a, 8, 5b, 2]`. `2` is swapped with `5a`, resulting in
                    `[2, 8, 5b, 5a]`. `5a` and `5b`'s relative order changed.
    """
    n = len(arr)
    # Traverse through all array elements
    for i in range(n):
        # Find the minimum element in remaining unsorted array
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j

        # Swap the found minimum element with the first element of the unsorted part
        # This places the minimum element at its correct sorted position
        arr[i], arr[min_idx] = arr[min_idx], arr[i]

    # The function modifies the array in-place.
```